{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp_hier</th>\n",
       "      <th>vista_id</th>\n",
       "      <th>allele_id</th>\n",
       "      <th>backbone</th>\n",
       "      <th>stage</th>\n",
       "      <th>curation_status</th>\n",
       "      <th>description</th>\n",
       "      <th>denominator</th>\n",
       "      <th>tissue</th>\n",
       "      <th>tissue_positive</th>\n",
       "      <th>...</th>\n",
       "      <th>coord</th>\n",
       "      <th>strand</th>\n",
       "      <th>coordinate_hg38</th>\n",
       "      <th>coordinate_mm10</th>\n",
       "      <th>bracket_hg38</th>\n",
       "      <th>bracket_mm10</th>\n",
       "      <th>seq_hg38</th>\n",
       "      <th>seq_mm10</th>\n",
       "      <th>seq</th>\n",
       "      <th>external_note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>03vn00010001</td>\n",
       "      <td>hs1</td>\n",
       "      <td>0</td>\n",
       "      <td>hZR</td>\n",
       "      <td>e11.5</td>\n",
       "      <td>positive</td>\n",
       "      <td>reference</td>\n",
       "      <td>15.0</td>\n",
       "      <td>lb;hb;nt;cn</td>\n",
       "      <td>3;12;12;8</td>\n",
       "      <td>...</td>\n",
       "      <td>chr16:86396481-86397120</td>\n",
       "      <td>+</td>\n",
       "      <td>chr16:86396481-86397120</td>\n",
       "      <td>chr8:121002895-121003512</td>\n",
       "      <td>IRF8-FOXF1</td>\n",
       "      <td>Gm20388 (intragenic)</td>\n",
       "      <td>AACTGAAGGGACCCCGTTAGCATATAAACAAAAGGTGGGGGGTAGC...</td>\n",
       "      <td>AGTTGAGCGACCCTGTTAACGTATAAACAAAAGGTGGGGGGTAACC...</td>\n",
       "      <td>AACTGAAGGGACCCCGTTAGCATATAAACAAAAGGTGGGGGGTAGC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>03v300010001</td>\n",
       "      <td>hs2</td>\n",
       "      <td>0</td>\n",
       "      <td>hZR</td>\n",
       "      <td>e11.5</td>\n",
       "      <td>negative</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chr16:85586489-85588130</td>\n",
       "      <td>+</td>\n",
       "      <td>chr16:85586489-85588130</td>\n",
       "      <td>chr8:120516455-120517853</td>\n",
       "      <td>GSE1 (intragenic)</td>\n",
       "      <td>Gm20388 (intragenic)</td>\n",
       "      <td>GGCCCTGGTATGTTTGTTCTTCCAGGGGCTCCCAGGATGGATCCAG...</td>\n",
       "      <td>ATATTGGCTCCTACAGGGGCTCCAGGATCATTCCCTGCCCTCCTGG...</td>\n",
       "      <td>GGCCCTGGTATGTTTGTTCTTCCAGGGGCTCCCAGGATGGATCCAG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>000100010001</td>\n",
       "      <td>hs3</td>\n",
       "      <td>0</td>\n",
       "      <td>hZR</td>\n",
       "      <td>e11.5</td>\n",
       "      <td>negative</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chr16:80389446-80390755</td>\n",
       "      <td>+</td>\n",
       "      <td>chr16:80389446-80390755</td>\n",
       "      <td>chr8:116384137-116385431</td>\n",
       "      <td>MAF-DYNLRB2</td>\n",
       "      <td>Maf-Dynlrb2</td>\n",
       "      <td>AAGATTGCCATTTGGGGTGTTTCTTGGGGCTAAGAACCATGAAGAC...</td>\n",
       "      <td>GAGATTGCCATTTGGGGTGTGTCTTGGGGGCTAAGTGCCATGAAGA...</td>\n",
       "      <td>AAGATTGCCATTTGGGGTGTTTCTTGGGGCTAAGAACCATGAAGAC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>000200010001</td>\n",
       "      <td>hs4</td>\n",
       "      <td>0</td>\n",
       "      <td>hZR</td>\n",
       "      <td>e11.5</td>\n",
       "      <td>positive</td>\n",
       "      <td>reference</td>\n",
       "      <td>10.0</td>\n",
       "      <td>mb;hb;nt</td>\n",
       "      <td>10;10;6</td>\n",
       "      <td>...</td>\n",
       "      <td>chr16:80338700-80339858</td>\n",
       "      <td>+</td>\n",
       "      <td>chr16:80338700-80339858</td>\n",
       "      <td>chr8:116335161-116337629</td>\n",
       "      <td>MAF-DYNLRB2</td>\n",
       "      <td>Maf-Dynlrb2</td>\n",
       "      <td>CAGAGACAGACAGTGACAGAGACAGATTTTAGAATTTGAACAAAGG...</td>\n",
       "      <td>catgtgtgtatgtgtgcatatgtgtgtgtgtgCTGGCGCACATTGA...</td>\n",
       "      <td>CAGAGACAGACAGTGACAGAGACAGATTTTAGAATTTGAACAAAGG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>03ya00010001</td>\n",
       "      <td>hs5</td>\n",
       "      <td>0</td>\n",
       "      <td>hZR</td>\n",
       "      <td>e11.5</td>\n",
       "      <td>negative</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chr16:79936010-79937400</td>\n",
       "      <td>+</td>\n",
       "      <td>chr16:79936010-79937400</td>\n",
       "      <td>chr8:116013572-116015153</td>\n",
       "      <td>MAF-DYNLRB2</td>\n",
       "      <td>Maf-Dynlrb2</td>\n",
       "      <td>TGACACCCACTATTATCCAGTCCTTGATAAACCTCTTTATTTGTTC...</td>\n",
       "      <td>TGTCTCCTACAGCTATAGTTCTTCAGTTATTGACAGAACAGACCCT...</td>\n",
       "      <td>TGACACCCACTATTATCCAGTCCTTGATAAACCTCTTTATTTGTTC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exp_hier vista_id  allele_id backbone  stage curation_status  \\\n",
       "0  03vn00010001      hs1          0      hZR  e11.5        positive   \n",
       "1  03v300010001      hs2          0      hZR  e11.5        negative   \n",
       "2  000100010001      hs3          0      hZR  e11.5        negative   \n",
       "3  000200010001      hs4          0      hZR  e11.5        positive   \n",
       "4  03ya00010001      hs5          0      hZR  e11.5        negative   \n",
       "\n",
       "  description  denominator       tissue tissue_positive  ...  \\\n",
       "0   reference         15.0  lb;hb;nt;cn       3;12;12;8  ...   \n",
       "1   reference          NaN          NaN             NaN  ...   \n",
       "2   reference          NaN          NaN             NaN  ...   \n",
       "3   reference         10.0     mb;hb;nt         10;10;6  ...   \n",
       "4   reference          NaN          NaN             NaN  ...   \n",
       "\n",
       "                     coord strand          coordinate_hg38  \\\n",
       "0  chr16:86396481-86397120      +  chr16:86396481-86397120   \n",
       "1  chr16:85586489-85588130      +  chr16:85586489-85588130   \n",
       "2  chr16:80389446-80390755      +  chr16:80389446-80390755   \n",
       "3  chr16:80338700-80339858      +  chr16:80338700-80339858   \n",
       "4  chr16:79936010-79937400      +  chr16:79936010-79937400   \n",
       "\n",
       "            coordinate_mm10       bracket_hg38          bracket_mm10  \\\n",
       "0  chr8:121002895-121003512         IRF8-FOXF1  Gm20388 (intragenic)   \n",
       "1  chr8:120516455-120517853  GSE1 (intragenic)  Gm20388 (intragenic)   \n",
       "2  chr8:116384137-116385431        MAF-DYNLRB2           Maf-Dynlrb2   \n",
       "3  chr8:116335161-116337629        MAF-DYNLRB2           Maf-Dynlrb2   \n",
       "4  chr8:116013572-116015153        MAF-DYNLRB2           Maf-Dynlrb2   \n",
       "\n",
       "                                            seq_hg38  \\\n",
       "0  AACTGAAGGGACCCCGTTAGCATATAAACAAAAGGTGGGGGGTAGC...   \n",
       "1  GGCCCTGGTATGTTTGTTCTTCCAGGGGCTCCCAGGATGGATCCAG...   \n",
       "2  AAGATTGCCATTTGGGGTGTTTCTTGGGGCTAAGAACCATGAAGAC...   \n",
       "3  CAGAGACAGACAGTGACAGAGACAGATTTTAGAATTTGAACAAAGG...   \n",
       "4  TGACACCCACTATTATCCAGTCCTTGATAAACCTCTTTATTTGTTC...   \n",
       "\n",
       "                                            seq_mm10  \\\n",
       "0  AGTTGAGCGACCCTGTTAACGTATAAACAAAAGGTGGGGGGTAACC...   \n",
       "1  ATATTGGCTCCTACAGGGGCTCCAGGATCATTCCCTGCCCTCCTGG...   \n",
       "2  GAGATTGCCATTTGGGGTGTGTCTTGGGGGCTAAGTGCCATGAAGA...   \n",
       "3  catgtgtgtatgtgtgcatatgtgtgtgtgtgCTGGCGCACATTGA...   \n",
       "4  TGTCTCCTACAGCTATAGTTCTTCAGTTATTGACAGAACAGACCCT...   \n",
       "\n",
       "                                                 seq external_note  \n",
       "0  AACTGAAGGGACCCCGTTAGCATATAAACAAAAGGTGGGGGGTAGC...           NaN  \n",
       "1  GGCCCTGGTATGTTTGTTCTTCCAGGGGCTCCCAGGATGGATCCAG...           NaN  \n",
       "2  AAGATTGCCATTTGGGGTGTTTCTTGGGGCTAAGAACCATGAAGAC...           NaN  \n",
       "3  CAGAGACAGACAGTGACAGAGACAGATTTTAGAATTTGAACAAAGG...           NaN  \n",
       "4  TGACACCCACTATTATCCAGTCCTTGATAAACCTCTTTATTTGTTC...           NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "experimentd_df = pd.read_csv(\"experiments.tsv\", delimiter=\"\\t\")\n",
    "experimentd_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2267\n"
     ]
    }
   ],
   "source": [
    "positive_df = experimentd_df.loc[\n",
    "    experimentd_df[\"curation_status\"] == \"positive\", [\"seq_hg38\"]\n",
    "]\n",
    "positive_df = positive_df[positive_df[\"seq_hg38\"].notna()]\n",
    "positive_df[\"class\"] = 1\n",
    "print(positive_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1913\n"
     ]
    }
   ],
   "source": [
    "\n",
    "negative_df = experimentd_df.loc[experimentd_df['curation_status'] == 'negative', ['seq_hg38']]\n",
    "negative_df = negative_df[negative_df['seq_hg38'].notna()]\n",
    "negative_df['class'] = 0\n",
    "print(negative_df.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from itertools import product\n",
    "\n",
    "\n",
    "def count_kmers(seqence, k):\n",
    "    nucleotides = [\"A\", \"C\", \"G\", \"T\"]\n",
    "    all_subsequences = [\"\".join(seq) for seq in product(nucleotides, repeat=k)]\n",
    "    k_mers = set([])\n",
    "    for k_mer in all_subsequences:\n",
    "        seq = Seq(k_mer)\n",
    "        if str(seq.reverse_complement()) not in k_mers:\n",
    "            k_mers.add(seq)\n",
    "\n",
    "    seq = Seq(seqence.upper())\n",
    "    k_mer_freq = {\n",
    "        key: (\n",
    "            seq.count_overlap(key)\n",
    "            + (\n",
    "                seq.count_overlap(key.reverse_complement())\n",
    "                if key != key.reverse_complement()\n",
    "                else 0\n",
    "            )\n",
    "        )\n",
    "        / len(seq)\n",
    "        for key in k_mers\n",
    "    }\n",
    "    return [_freq for _seq, _freq in sorted(k_mer_freq.items())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_data(src_df, k):\n",
    "    df = pd.DataFrame()\n",
    "    df[\"class\"] = src_df[\"class\"]\n",
    "    features_len = len(count_kmers(\"_\", k))\n",
    "    features_df = src_df[\"seq_hg38\"].apply(lambda x: pd.Series(count_kmers(x, k)))\n",
    "    features_df.columns = [\"x\" + str(i) for i in range(features_len)]\n",
    "\n",
    "    df = pd.concat([df, features_df], axis=1)\n",
    "    X = df.drop(columns=[\"class\"])\n",
    "    Y = df.loc[:, \"class\"]\n",
    "    return X, Y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.linear_model import LogisticRegressionCV\n",
    "from sklearn.metrics import (\n",
    "    accuracy_score,\n",
    "    recall_score,\n",
    "    precision_score,\n",
    "    roc_auc_score,\n",
    "    f1_score,\n",
    "    confusion_matrix,\n",
    ")\n",
    "\n",
    "\n",
    "def evaluate_model(y_pred, y_proba, y_true):\n",
    "    accuracy = accuracy_score(y_true, y_pred)\n",
    "    precision = precision_score(y_true, y_pred)\n",
    "    recall = recall_score(y_true, y_pred)\n",
    "    roc_auc = roc_auc_score(y_true, y_proba)\n",
    "    f1 = f1_score(y_true, y_pred)\n",
    "    conf_matrix = confusion_matrix(y_true, y_pred)\n",
    "\n",
    "    return {\n",
    "        \"accuracy\": accuracy,\n",
    "        \"precision\": precision,\n",
    "        \"recall\": recall,\n",
    "        \"roc_auc\": roc_auc,\n",
    "        \"f1_score\": f1,\n",
    "    }, conf_matrix\n",
    "\n",
    "\n",
    "def train_and_evaluate_model(pos_df, neg_df, k):\n",
    "    train_pos, test_pos = train_test_split(pos_df, test_size=0.3, random_state=0)\n",
    "    train_neg, test_neg = train_test_split(neg_df, test_size=0.3, random_state=0)\n",
    "    X_train, Y_train = preprocess_data(pd.concat([train_pos, train_neg], axis=0), k)\n",
    "    X_test, Y_test = preprocess_data(pd.concat([test_pos, test_neg], axis=0), k)\n",
    "\n",
    "    model = LogisticRegressionCV(cv=10, random_state=0, n_jobs=-1)\n",
    "    model.fit(X_train, Y_train)\n",
    "    cv_mean = model.scores_[1].mean(axis=0)\n",
    "    cv_std = model.scores_[1].std(axis=0)\n",
    "    max_cv_mean_i = cv_mean.argmax()\n",
    "    y_pred = model.predict(X_test)\n",
    "    y_proba = model.predict_proba(X_test)[:, 1]\n",
    "    metrics, matrices = evaluate_model(y_pred, y_proba, Y_test)\n",
    "    metrics[\"cv_accuracy\"] = cv_mean[max_cv_mean_i]\n",
    "    metrics[\"cv_accuracy_std\"] = cv_std[max_cv_mean_i]\n",
    "    return metrics, matrices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accuracy</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>roc_auc</th>\n",
       "      <th>f1_score</th>\n",
       "      <th>cv_accuracy</th>\n",
       "      <th>cv_accuracy_std</th>\n",
       "      <th>k</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.564940</td>\n",
       "      <td>0.576792</td>\n",
       "      <td>0.744493</td>\n",
       "      <td>0.584744</td>\n",
       "      <td>0.650000</td>\n",
       "      <td>0.570596</td>\n",
       "      <td>0.023293</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.576096</td>\n",
       "      <td>0.585338</td>\n",
       "      <td>0.750367</td>\n",
       "      <td>0.587717</td>\n",
       "      <td>0.657658</td>\n",
       "      <td>0.582572</td>\n",
       "      <td>0.027811</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.556972</td>\n",
       "      <td>0.555457</td>\n",
       "      <td>0.919236</td>\n",
       "      <td>0.579126</td>\n",
       "      <td>0.692478</td>\n",
       "      <td>0.564098</td>\n",
       "      <td>0.010428</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   accuracy  precision    recall   roc_auc  f1_score  cv_accuracy  \\\n",
       "0  0.564940   0.576792  0.744493  0.584744  0.650000     0.570596   \n",
       "1  0.576096   0.585338  0.750367  0.587717  0.657658     0.582572   \n",
       "2  0.556972   0.555457  0.919236  0.579126  0.692478     0.564098   \n",
       "\n",
       "   cv_accuracy_std  k  \n",
       "0         0.023293  3  \n",
       "1         0.027811  4  \n",
       "2         0.010428  5  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scores_list = []\n",
    "conf_matrices = []\n",
    "for k in [3, 4, 5]:\n",
    "    metrics, conf_matrix = train_and_evaluate_model(positive_df, negative_df, k)\n",
    "    metrics[\"k\"] = k\n",
    "    scores_list.append(metrics)\n",
    "    conf_matrices.append(conf_matrix)\n",
    "scores_df = pd.DataFrame(scores_list)\n",
    "scores_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[212, 362],\n",
       "       [170, 511]])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conf_matrices[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Random negative data from GRCh38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from Bio import SeqIO\n",
    "\n",
    "sample_sequences = {} \n",
    "for record in SeqIO.parse('GRCh38.p14.genome.fa', format='fasta'):\n",
    "    sample_sequences[record.id] = record.seq\n",
    "    if len(sample_sequences) == 23: break\n",
    "\n",
    "seq_lengths = positive_df['seq_hg38'].apply(len).tolist()\n",
    "random.seed(0)\n",
    "positive_sequences = set(positive_df['seq_hg38'].apply(str.upper).tolist())\n",
    "\n",
    "positive_coordinates_df = experimentd_df.loc[experimentd_df['curation_status'] == 'positive', ['coordinate_hg38']]\n",
    "positive_coordinates_df = positive_coordinates_df[positive_coordinates_df['coordinate_hg38'].notna()]\n",
    "positive_coordinates_df['chromosome'] = positive_coordinates_df['coordinate_hg38'].apply(lambda x: x.split(':')[0])\n",
    "positive_coordinates_df['range'] = positive_coordinates_df['coordinate_hg38'].apply(lambda x: [int(item) for item in x.split(':')[1].split('-')])\n",
    "positive_coordinastes = positive_coordinates_df.groupby('chromosome')['range'].apply(list).to_dict()\n",
    "\n",
    "negative_sequences = []\n",
    "for length in seq_lengths:\n",
    "    sample_chromosome = random.choice(list(sample_sequences.keys()))\n",
    "    sample_sequence = sample_sequences[sample_chromosome]\n",
    "    coordinates_to_exclude = positive_coordinastes[sample_chromosome]\n",
    "    start = random.randint(0,len(sample_sequence) - length)\n",
    "    end_excl = start + length\n",
    "    sequence = sample_sequence[start:end_excl]\n",
    "    while sequence.count(\"N\") > 0 or any(\n",
    "        [\n",
    "            start >= coord[0] and start <= coord[1] or end_excl - 1 >= coord[0] and end_excl -1 <= coord[1]\n",
    "            for coord in coordinates_to_exclude\n",
    "        ]\n",
    "    ):\n",
    "        start = random.randint(0, len(sample_sequence) - length)\n",
    "        end_excl = start + length\n",
    "        sequence = sample_sequence[start:end_excl]\n",
    "    negative_sequences.append(str(sequence).upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "artificial_negative_df = pd.DataFrame()\n",
    "artificial_negative_df[\"seq_hg38\"] = negative_sequences\n",
    "artificial_negative_df[\"class\"] = 0\n",
    "\n",
    "artificial_positive_df = pd.DataFrame()\n",
    "artificial_positive_df[\"seq_hg38\"] = positive_df[\"seq_hg38\"]\n",
    "artificial_positive_df[\"class\"] = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accuracy</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>roc_auc</th>\n",
       "      <th>f1_score</th>\n",
       "      <th>cv_accuracy</th>\n",
       "      <th>cv_accuracy_std</th>\n",
       "      <th>k</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.751101</td>\n",
       "      <td>0.752212</td>\n",
       "      <td>0.748899</td>\n",
       "      <td>0.815483</td>\n",
       "      <td>0.750552</td>\n",
       "      <td>0.747168</td>\n",
       "      <td>0.018725</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.779001</td>\n",
       "      <td>0.790520</td>\n",
       "      <td>0.759178</td>\n",
       "      <td>0.844424</td>\n",
       "      <td>0.774532</td>\n",
       "      <td>0.776164</td>\n",
       "      <td>0.020641</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.792217</td>\n",
       "      <td>0.794379</td>\n",
       "      <td>0.788546</td>\n",
       "      <td>0.854701</td>\n",
       "      <td>0.791452</td>\n",
       "      <td>0.780583</td>\n",
       "      <td>0.022453</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   accuracy  precision    recall   roc_auc  f1_score  cv_accuracy  \\\n",
       "0  0.751101   0.752212  0.748899  0.815483  0.750552     0.747168   \n",
       "1  0.779001   0.790520  0.759178  0.844424  0.774532     0.776164   \n",
       "2  0.792217   0.794379  0.788546  0.854701  0.791452     0.780583   \n",
       "\n",
       "   cv_accuracy_std  k  \n",
       "0         0.018725  3  \n",
       "1         0.020641  4  \n",
       "2         0.022453  5  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scores_list_a = []\n",
    "conf_matrices_a = []\n",
    "for k in [3, 4, 5]:\n",
    "    metrics, conf_matrix = train_and_evaluate_model(\n",
    "        artificial_positive_df, artificial_negative_df, k\n",
    "    )\n",
    "    metrics[\"k\"] = k\n",
    "    scores_list_a.append(metrics)\n",
    "    conf_matrices_a.append(conf_matrix)\n",
    "scores_df_a = pd.DataFrame(scores_list_a)\n",
    "scores_df_a.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[542, 139],\n",
       "       [144, 537]])"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conf_matrices_a[2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluate model trained with artificial negative on vista negative data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of test data: positive 681, negative 574\n",
      "[[182 392]\n",
      " [164 517]]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accuracy</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>roc_auc</th>\n",
       "      <th>f1_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.556972</td>\n",
       "      <td>0.568757</td>\n",
       "      <td>0.759178</td>\n",
       "      <td>0.554744</td>\n",
       "      <td>0.650314</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   accuracy  precision    recall   roc_auc  f1_score\n",
       "0  0.556972   0.568757  0.759178  0.554744  0.650314"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "k = 4\n",
    "train_pos, test_pos = train_test_split(\n",
    "    artificial_positive_df, test_size=0.3, random_state=0\n",
    ")\n",
    "train_neg, _ = train_test_split(\n",
    "    artificial_negative_df, test_size=0.3, random_state=0\n",
    ")\n",
    "_, test_neg = train_test_split(\n",
    "    negative_df, test_size=0.3, random_state=0\n",
    ")\n",
    "X_train, Y_train = preprocess_data(pd.concat([train_pos, train_neg], axis=0), k)\n",
    "\n",
    "\n",
    "print(f'Size of test data: positive {test_pos.shape[0]}, negative {test_neg.shape[0]}')\n",
    "model = LogisticRegressionCV(cv=10, random_state=0, n_jobs=-1)\n",
    "model.fit(X_train, Y_train)\n",
    "X_test, Y_test = preprocess_data(pd.concat([test_neg, test_pos], axis=0), k)\n",
    "y_pred = model.predict(X_test)\n",
    "y_proba = model.predict_proba(X_test)[:, 1]\n",
    "metrics, conf_matrix = evaluate_model(y_pred, y_proba, Y_test)\n",
    "a_eval_df = pd.DataFrame([metrics])\n",
    "print(conf_matrix)\n",
    "a_eval_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluate model trained with vista data on artificial negative data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of test data: positive 681, negative 681\n",
      "[[424 257]\n",
      " [170 511]]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accuracy</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>roc_auc</th>\n",
       "      <th>f1_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.68649</td>\n",
       "      <td>0.665365</td>\n",
       "      <td>0.750367</td>\n",
       "      <td>0.746939</td>\n",
       "      <td>0.705314</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   accuracy  precision    recall   roc_auc  f1_score\n",
       "0   0.68649   0.665365  0.750367  0.746939  0.705314"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "k = 4\n",
    "train_pos, test_pos = train_test_split(positive_df, test_size=0.3, random_state=0)\n",
    "train_neg, _ = train_test_split(negative_df, test_size=0.3, random_state=0)\n",
    "_, test_neg = train_test_split(artificial_negative_df, test_size=0.3, random_state=0)\n",
    "X_train, Y_train = preprocess_data(pd.concat([train_pos, train_neg], axis=0), k)\n",
    "print(f'Size of test data: positive {test_pos.shape[0]}, negative {test_neg.shape[0]}')\n",
    "\n",
    "model = LogisticRegressionCV(cv=10, random_state=0, n_jobs=-1)\n",
    "model.fit(X_train, Y_train)\n",
    "X_test, Y_test = preprocess_data(\n",
    "    pd.concat([test_pos, test_neg], axis=0), k\n",
    ")\n",
    "y_pred = model.predict(X_test)\n",
    "y_proba = model.predict_proba(X_test)[:, 1]\n",
    "metrics, conf_matrix = evaluate_model(y_pred, y_proba, Y_test)\n",
    "a2_eval_df = pd.DataFrame([metrics]).head()\n",
    "print(conf_matrix)\n",
    "a2_eval_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
